### PR TITLE
[WIP][runtime] Add circular reference check to create_runtime_class

### PR DESCRIPTION
**Pull request by alexanderkyte**
FAKE was running into a stack overflow due to a circular reference between the root TypeBuilder and the TypeBuilder (tbuilder) in the GenericTypeParameterBuilder in one of the subtypes of the root TypeBuilder.  ```root->subtypes[i]->generic_arguments[j]->tbuilder == root```

This is a runtime check for circular references passed to SRE, it should make debugging this kind of API misuse easier.

### CODE REVIEW

File: mono/metadata/sre.c [hunk 1]
```diff
@@ -2350,14 +2350,36 @@ mono_reflection_get_custom_attrs_blob_checked (MonoReflectionAssembly *assembly,
...
+	gpointer raw;
+
 	if (!MONO_HANDLE_IS_NULL (ref_parent)) {
+		raw = MONO_HANDLE_RAW (ref_parent);
+
+		gpointer in = g_hash_table_lookup (seen, raw);
+		if (in)
+			goto leave_circular;
+		else
+			g_hash_table_insert (seen, raw, GINT_TO_POINTER (TRUE));
```
Comment (kumpera) [inline mono/metadata/sre.c]: You can't insert a raw pointer into a GHashTable

File: mono/metadata/sre.c [hunk 2]
```diff
@@ -2350,14 +2350,36 @@ mono_reflection_get_custom_attrs_blob_checked (MonoReflectionAssembly *assembly,
 static gboolean
 reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoError *error)
 {
+	static GHashTable *seen = NULL;
```
Comment (kumpera) [inline mono/metadata/sre.c]: Making this static only works as long as we don't change anything in locking.

This is a bad design.

Comment (kumpera) [inline mono/metadata/sre.c]: I think we can fix this by renaming reflection_setup_internal_class to reflection_setup_internal_class_internal while adding a new function named reflection_setup_internal_class.

We'd do something like this: ```
static gboolean
reflection_setup_internal_class (MonoReflectionTypeBuilderHandle ref_tb, MonoError *error)
{
   MonoPtrArray types;
   mono_ptr_array_init (types, `6, MONO_ROOT_SOURCE_REFLECTION, "sre setup internal class");
   gboolean res = reflection_setup_internal_class_internal (ref_tb, &types, error);
   mono_ptr_array_destroy (types);
   return res;
}
```

And in reflection_setup_internal_class_internal we'd search `types` for what we're init'ing and add to it if not found. The average depth will never be big enough for this to matter.

**Comment by vargaz** [comment]
Its ok to have circular references between type builders.

**Pull request closed without merging**